docs/site/Express-middleware.md

```diff
@@ -0,0 +1,229 @@
+ * @param middlewareConfig - Middleware config
+ * @param options - Options for registration
+ *
+ * @typeParam CFG - Configuration type
+ */
+ middleware<CFG>(
+   middleware: string | MiddlewareFactory<CFG>,
+   middlewareConfig?: CFG,
+   options: MiddlewareActionBindingOptions = {},
+ ): Binding<RequestInterceptor> {
+   // ...
+ }
+````
+
+```ts
+const binding = app.middleware(spy, spyConfig, {
+  injectConfiguration: true,
+  key: 'interceptors.middleware.spy',
+});
+```
+
+#### Default sequence
+
+We add a new `middleware` property to `DefaultSequence` to allow injection of
+the middleware action while keeping backward compatibility.
+
+```ts
+export class DefaultSequence implements SequenceHandler {
+  /**
+   * Optional middleware chain
+   * Invokes registered middleware (injected via SequenceActions.MIDDLEWARE).
+   */
+  @inject(SequenceActions.MIDDLEWARE, {optional: true})
+  protected middleware?: Middleware;
+
+  // ...
+  async handle(context: RequestContext): Promise<void> {
+    try {
+      const {request, response} = context;
+      if (this.middleware != null) {
+        await this.middleware(context);
+      }
+      const route = this.findRoute(request);
+      const args = await this.parseParams(request, route);
+      const result = await this.invoke(route, args);
+
+      debug('%s result -', route.describe(), result);
+      this.send(response, result);
+    } catch (error) {
+      this.reject(context, error);
+    }
+  }
+}
+```
+
+#### Extend sequence with more than one middleware steps
+
+Sometimes we want to add middleware between built-in steps, for example, do some
+post-processing before the result is sent to the HTTP response.
+
+```ts
+class SequenceWithMiddleware extends DefaultSequence {
+  /**
+   * Optional middleware chain
+   * Invokes registered middleware (injected via SequenceActions.MIDDLEWARE).
+   */
+  @inject('middleware.afterInvoke', {optional: true})
+  protected middlewareAfterInvoke?: Middleware;
+  async handle(context: RequestContext): Promise<void> {
+    try {
+      const {request, response} = context;
+      if (this.middleware != null) {
+        await this.middleware(context);
+      }
+      const route = this.findRoute(request);
+      const args = await this.parseParams(request, route);
+      const result = await this.invoke(route, args);
+      // Invoke middleware after invoke
+      context.bind('invocation.result').to(result);
+      if (this.middlewareAfterInvoke != null) {
+        await this.middlewareAfterInvoke(context);
+      }
+      this.send(response, result);
+    } catch (error) {
+      this.reject(context, error);
+    }
+  }
+}
+```
+
+Now we can set up applications to leverage the new sequence:
+
+```ts
+// Create another middleware phase
+app
+  .bind('middleware.afterInvoke')
+  .toProvider(MiddlewareProvider)
+  .tag({[CoreTags.EXTENSION_POINT]: POST_INVOCATION_MIDDLEWARE});
+helper.app.sequence(SequenceWithMiddleware);
```

Comment:
What is `helper`? Where does it come from? Can't we just use `app.sequence(SequenceWithMiddleware)`?